### ISSUE TITLE
Add first-class ngrok support

### ISSUE THREAD

**Issue by rjygraham** [dotnet/tye#980]
Add first-class ngrok support as an ingress controller so requests from remote devices can invoke services running on the local dev machine.

This is important to me because one of the pain points developing Android/iOS apps is accessing APIs hosted on the local dev machine. By incorporating ngrok support in Tye, this pain would be completely removed and worked into the normal Tye development workflow.

Today, ngrok can be added to Tye via an `executable`; however, the ngrok output is not captured by Tye and therefore the developer won't know the subdomain automatically generated for their proxy session. To get around this, the ngrok output can be logged to an external file or the developer can discover it via invoking the ngrok api: `http://localhost:4040/api/tunnels`.

**Comment by codeputer** [comment]
Best product.... https://github.com/OfficeDev/microsoft-teams-tunnelrelay

Let know what you think of this tool

R

**Comment by rjygraham** [comment]
There are are couple things to love about TunnelRelay - flexibility via plugins, the fact it goes over a TLS channel that you control, etc. However, it's minimum $10/mo use since you have to bring your own Hybrid Connection.

On the other hand, ngrok is completely free to get started...

Regardless, thank you for pointing me to TunnelRelay as I've contemplated building this exact thing several times in the past. Instead of from scratch I can contribute to this existing tool.

**Comment by Misiu** [comment]
> On the other hand, ngrok is completely free to get started...

I totally agree. I used ngrok in the bast to develop and debug webhooks solutions.
The main advantage is that it is totally free to get started.

So we would need to catch the output of ngrokand parse it to get the subdomain.
Maybe the easiest way would be to open `http://127.0.0.1:4040/` in the browser after ngrok starts?

**Comment by codeputer** [comment]
I've used both...and recently invoiced a client show them how to use an Azure Relay Service... I also forked this to picandpost (PR wasn't accepted) something I've done multiple times😆

Ha


I have a lot of fun being hired to show how internal access, and a tool like this, can route in requests to a local work station, by passing the 10 day build process to Dev which has a public point that can route in from public third party services, and attach a debugger.

I would appreciate a 20min teams coffee? This link gets into my cloud based contact sms

https://callrichard.direct/

Regards,
Richard


________________________________

**Comment by Misiu** [comment]
Another alternative is Cloudflare Argo Tunnel (https://www.cloudflare.com/products/tunnel/).
